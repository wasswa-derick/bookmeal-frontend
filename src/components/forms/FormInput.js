import React from "react";
import PropTypes from "prop-types";
import InlineError from "../InlineError";

const FormInput = ({ value, type, name, label, onChange, error }) => (
  <div className="form-group">
    <label htmlFor={name}>{label}</label>
    <input
      className={error ? "form-control is-invalid" : "form-control"}
      type={type}
      id={name}
      autoComplete="off"
      value={value}
      name={name}
      onChange={onChange}
    />
    {error && <InlineError text={error} />}
  </div>
);

FormInput.propTypes = {
  value: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  error: PropTypes.string
};

FormInput.defaultProps = {
  error: undefined
};

export default FormInput;
