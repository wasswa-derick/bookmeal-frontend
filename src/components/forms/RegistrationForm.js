import React from "react";
import validator from "validator";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import FormInput from "./FormInput";

/**
 * @class RegistrationForm
 * @extends {React.Component}
 */
class RegistrationForm extends React.Component {
  state = {
    data: {
      name: "",
      email: "",
      password: ""
    },
    errors: {}
  };

  /**
   * @param {Event} evt
   *@returns {null} sets value of target element to the corresponding state value
   */
  onChange = evt => {
    this.setState({
      data: { ...this.state.data, [evt.target.name]: evt.target.value }
    });
  };
  /**
   *@param {Event} evt
   * @returns {null} returns null
   */
  onFormSubmit = evt => {
    evt.preventDefault();
    const { data } = this.state;
    let errors = this.validateData(data);

    if (Object.keys(errors).length === 0) {
      this.props.handleSubmit(data).catch(err => {
        switch (err.response.status) {
          case 500:
            errors.form = "Internal server error occured";
            break;
          case 400:
            errors = { ...err.response.data.errors };
            break;
          default:
            break;
        }
        this.setState({ errors });
      });
    }

    this.setState({ errors });
  };
  /**
   *@param {Object} data
   * @returns {Object} errors object
   */
  validateData = data => {
    const errors = {};
    if (validator.isEmpty(data.email)) {
      errors.email = "This field is required";
    } else if (!validator.isEmail(data.email)) {
      errors.email = "This email is invalid";
    }
    if (validator.isEmpty(data.password)) {
      errors.password = "This field is required";
    }
    if (validator.isEmpty(data.name)) {
      errors.name = "This field is required";
    }
    return errors;
  };

  /**
   * @returns {form} for registration
   */
  render() {
    const { data, errors } = this.state;
    return (
      <form autoComplete="off" onSubmit={this.onFormSubmit}>
        <FormInput
          value={data.name}
          type="text"
          name="name"
          label="Full Name"
          onChange={this.onChange}
          error={errors.name}
        />
        <FormInput
          name="email"
          type="email"
          label="Email"
          value={data.email}
          onChange={this.onChange}
          error={errors.email}
        />
        <FormInput
          name="password"
          type="password"
          label="Password"
          value={data.password}
          onChange={this.onChange}
          error={errors.password}
        />
        <button type="submit" className="btn btn-primary btn-submit">
          Create Account
        </button>
        <p>
          Already have an account?
          <Link href="/login" to="/login">
            Log In
          </Link>
        </p>
      </form>
    );
  }
}

RegistrationForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired
};

export default RegistrationForm;
