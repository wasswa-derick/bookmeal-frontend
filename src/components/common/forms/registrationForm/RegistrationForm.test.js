import React from "react";
import { shallow } from "enzyme";
import RegistrationForm from "./RegistrationForm";
import FormInput from "../formInput/FormInput";

describe("RegistrationForm", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<RegistrationForm handleSubmit={jest.fn} />);
  });

  it("should render form", () => {
    expect(wrapper.find("form").length).toEqual(1);
  });

  it("should render the FormInput component for Name", () => {
    expect(
      wrapper.containsMatchingElement(
        <FormInput
          onChange={wrapper.instance().onChange}
          type="text"
          name="name"
          label="Full Name"
          value=""
        />
      )
    ).toEqual(true);
  });

  it("should render a link to login", () => {
    const linkProps = wrapper.find("Link").props();
    expect(linkProps.to).toBe("/login");
  });

  it("should change state on change of target value", () => {
    const evt = {
      target: {
        name: "password",
        value: "password1"
      }
    };

    wrapper.instance().onChange(evt);
    const { data } = wrapper.instance().state;
    expect(data.password).toBe("password1");
  });

  it("it calls onSubmit when form is submitted", () => {
    const spy = jest.spyOn(wrapper.instance(), "onFormSubmit");
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);
  });

  it("should validate on submitting form", () => {
    wrapper.instance().onFormSubmit({ preventDefault: jest.fn });
    const { errors } = wrapper.instance().state;
    expect(errors.email).toBe("This field is required");
    expect(errors.password).toBe("This field is required");
    expect(errors.name).toBe("This field is required");
  });

  it("should validate email on submit", () => {
    const evt = {
      target: { name: "email", value: "test@" }
    };

    wrapper.instance().onChange(evt);
    wrapper.instance().onFormSubmit({ preventDefault: jest.fn });
    const { errors } = wrapper.instance().state;
    expect(errors.email).toBe("This email is invalid");
  });

  it("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
