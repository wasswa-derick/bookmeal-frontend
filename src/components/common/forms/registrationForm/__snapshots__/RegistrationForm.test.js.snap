// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm should render correctly 1`] = `
<form
  autoComplete="off"
  onSubmit={[Function]}
>
  <FormInput
    label="Full Name"
    name="name"
    onChange={[Function]}
    type="text"
    value=""
  />
  <FormInput
    label="Email"
    name="email"
    onChange={[Function]}
    type="email"
    value=""
  />
  <FormInput
    label="Password"
    name="password"
    onChange={[Function]}
    type="password"
    value=""
  />
  <button
    className="btn btn-primary btn-submit"
    type="submit"
  >
    Create Account
  </button>
  <p>
    Already have an account?
    <Link
      href="/login"
      replace={false}
      to="/login"
    >
      Log In
    </Link>
  </p>
</form>
`;
